mktrayicon.c

```diff
@@ -256,10 +300,99 @@ gpointer watch_fifo(gpointer argv)
 #endif
 			break;
 		case 'm': /* menu */
-			fprintf(stderr, "Menu command not yet implemented\n");
-			if (param != NULL) {
-				free(param);
+			if (onmenu != NULL){
+				//destroy the previous menu
+				for (int i = 0; i < menusize; i++){
+					free(onmenu[i].name);
+					free(onmenu[i].action);
+					onmenu[i].name = NULL;
+					onmenu[i].action = NULL;
+				}
+				free(onmenu);
+				onmenu = NULL;
+				gtk_widget_destroy(menu);
+				menu = NULL;
 			}
+			
+			menusize = 0;
+
+			if (param != NULL && *param == '\0') {
+#ifdef DEBUG
+				printf("Removing onmenu handler\n");
+#endif
+				break;
+			}
+
+			// This block makes sure that the parameter after 'm' is well defined
+			int commas = 0;
+			int bars = 0;
+			if (len < 3){
+				break;
+			}
+			if (param[0] == ',' || param[0] == '|'){
+				printf("Action string cannot begin with ',' or '|'. Use '\\' to escape\n");
+				break;
+			}
+			if (param[len-2] == '|' && param[len-3] != '\\'){
+				printf("'|' can't be second of last. Use '\\' to escape\n");
+				break;
+			}
+			if (param[len-1] == ',' && param[len-2] != '\\'){
+				printf("',' can't be last. Use '\\' to escape\n");
+				break;
+			}
+			int i = 1; 
+			while (i < len - 1){
+				if (commas > bars + 1 || bars > commas)
+					break;
+				if (param[i] == ',' && param[i-1] != '\\')
+					commas++;
+				else if (param[i] == '|' && param[i-1] != '\\')
+					bars++;
+				i++;
+			}
+			if (commas > bars + 1 || bars > commas){
+				printf("Unbalanced ',' or '|'. Use '\\' to escape\n");
+				break;
+			}
+			// End of block that checks the parameter
+			
+			// Create the onmenu array which stores structs with name, action properties
+			menusize = commas;
+			onmenu = malloc(menusize * sizeof(struct item));
+			menu = gtk_menu_new();
+			int last = -1;
+			int item = 0;
+			i = 1;
```

Comment:
Shouldn't this start at `[0]`?

Reply:
If you mean `last` no because of the index arithmetic I use below. It's a bit confusing I know. I may have to clear that up. The same applies for `i` since I have forbidden the 0 index to be one of the delimiter characters from the code above, so no use in starting from 0. Anyway, I can look to clear that up.